core/src/main/scala/com/softwaremill/session/JwsAlgorithm.scala

```diff
@@ -0,0 +1,67 @@
+package com.softwaremill.session
+
+import java.nio.charset.StandardCharsets.UTF_8
+import java.security.spec.PKCS8EncodedKeySpec
+import java.security.{KeyFactory, PrivateKey, Signature}
+import java.util.{Base64, NoSuchElementException}
+
+import com.typesafe.config.Config
+
+import scala.util.{Failure, Try}
+
+sealed trait JwsAlgorithm {
+  def sign(message: String): String
+}
+
+object JwsAlgorithm {
+
+  case class Rsa(privateKey: PrivateKey) extends JwsAlgorithm {
+
+    override def sign(message: String): String = {
+      val privateSignature: Signature = Signature.getInstance("SHA256withRSA")
+      privateSignature.initSign(privateKey)
+      privateSignature.update(message.getBytes(UTF_8))
+
+      Base64.getEncoder.encodeToString(privateSignature.sign())
+    }
+  }
+
+  object Rsa {
+
+    def fromConfig(jwsConfig: Config): Try[Rsa] = {
+
+      def readKeyFromConf(): Try[String] = {
+        Try {
+          val configKey = "rsa-private-key"
+          Option(jwsConfig.hasPath(configKey))
+            .filter(identity)
+            .flatMap(_ => Option(jwsConfig.getString(configKey)))
+            .filter(_.trim.nonEmpty)
+            .map(_.replaceAll("\\s", "").replaceAll("-----[^-]+-----", ""))
+            .get
```

Comment:
first invoking `.get` to get a `NoSuchElementException` and then catching this looks a bit weird. Why not just putting `.getOrElse` here with the `IllegalArgumentException`, and moving the `.recoverWith` to the `Try` which tries to parse the key? That way the error handling is closer to where the errors can occur.,